internal/dot/graph.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dot
+
+// Node represents a single node in a graph.
+type Node struct {
+	Type     string
+	Name     string
+	Optional bool
+	Group    string
+}
+
+// Edge connects a node parameter to a node result.
+type Edge struct {
+	Param  *Node
+	Result *Node
+}
+
+// Graph represents the DOT-format graph in a Container.
+type Graph struct {
+	Edges []*Edge
+}
+
+// Add adds the edges in node n into dg.
+func (dg *Graph) Add(params []*Node, results []*Node) {
```

Comment:
Related to our offline discussion: This is basically `Add(constructor details)`
so it can be,

```go
type Constructor struct {
    Params  []*Node
    Results []*Node
}

func (dg *Graph) Add(*Constructor) {
    // ...
}
```

Reply:
Yes, the structure of the graph will be changed in the next PR after this is merged